### ISSUE TITLE
edgeAgent fails to load: 401 Unauthorized

### ISSUE THREAD

**Issue by jlorich** [Azure/iotedge#4049]
## Overview
I am working with a customer to deploy IoT Edge inside their factory.  They have a requirement that all traffic flow over port 443 and through a corporate proxy.  They're using X.509 certificates with a DPS Group Enrollment.  The Edge box is running RHEL 7.8 and Edge `1.0.10.2-1` is installed.

All 3 proxy locations are set correctly, AmqpWs is configured and Edge successfully completes registration with DPS and IoT Hub.

## Expected Behavior
The device should be able to start EdgeAgent and come online.

## Current Behavior
EdgeAgent fails to start and a 401 is shown in the logs.

## Steps to Reproduce

Outside of the customer environment I have not been able to reproduce this issue.  Here are things that I've tried:

- Intentionally misconfigured docker proxy URLs
- Intentionally misconfigured edge daemon proxy url
- Intentionally misconfigured edge agent/hub proxy url
- Intentionally misconfigured system time and hwclock time
- Intentionally blocking AMQP over TCP and turning of AmqpWs
- Deleting the device from IoT Hub but not DPS and restarting edge with `always_reprovision_on_startup` and `dynamic_reprovisioning` off

Each of these produces failures (except changing the time, which does nothing unless I'm using SAS tokens), but nothing that matches what my customer sees: DPS succeeding yey still getting a 401 from IoT hub.

I am also able to use *their* certificates on *my* Edge box to connect to *their* DPS and Hub and it works without issue.

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```
Configuration checks
--------------------
√ config.yaml is well-formed - OK
√ config.yaml has well-formed connection string - OK
√ container engine is installed and functional - OK
√ config.yaml has correct hostname - OK
√ config.yaml has correct URIs for daemon mgmt endpoint - OK
√ latest security daemon - OK
‼ host time is close to real time - Warning
    Could not query NTP server
√ container time is close to host time - OK
‼ DNS server - Warning
    Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.
    Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.
    You can ignore this warning if you are setting DNS server per module in the Edge deployment.
√ production readiness: identity certificates expiry - OK
‼ production readiness: certificates - Warning
    The Edge device is using self-signed automatically-generated development certificates.
    They will expire in 89 days (at 2021-03-04 10:27:22 UTC) causing module-to-module and downstream device communication to fail on an active deployment.
    After the certs have expired, restarting the IoT Edge daemon will trigger it to generate new development certs.
    Please consider using production certificates instead. See https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: container engine - OK
‼ production readiness: logs policy - Warning
    Container engine is not configured to rotate module logs which may cause it run out of disk space.
    Please see https://aka.ms/iotedge-prod-checklist-logs for best practices.
    You can ignore this warning if you are setting log policy per module in the Edge deployment.
× production readiness: Edge Agent's storage directory is persisted on the host filesystem - Error
    Could not check current state of edgeAgent container
× production readiness: Edge Hub's storage directory is persisted on the host filesystem - Error
    Could not check current state of edgeHub container
 
Connectivity checks
-------------------
× host can connect to and perform TLS handshake with DPS endpoint - Error
    Could not connect: Connection reset by peer (os error 104)
√ host can connect to and perform TLS handshake with IoT Hub AMQP port - OK
√ host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - OK
√ host can connect to and perform TLS handshake with IoT Hub MQTT port - OK
√ container on the default network can connect to IoT Hub AMQP port - OK
√ container on the default network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the default network can connect to IoT Hub MQTT port - OK
√ container on the IoT Edge module network can connect to IoT Hub AMQP port - OK
√ container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - OK
√ container on the IoT Edge module network can connect to IoT Hub MQTT port - OK
 
18 check(s) succeeded.
4 check(s) raised warnings. Re-run with --verbose for more details.
3 check(s) raised errors. Re-run with --verbose for more details.


```
</details>

### Device Information
* Host OS: RHEL 7.8 
* Architecture: amd64 
* Container OS: Linux containers 

### Runtime Versions
* iotedged 1.0.10.2: 
* Edge Agent 1.0.0: 
* Edge Hub 1.0.0: 

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs

<details>
<summary>iotedged logs</summary>

```
Z [INFO] - Starting DPS registration with scope_id "0ne0019F2DB", registration_id "XXXXXXXXXX"
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - DPS registration assigned device XXXXXXXXXXin hub XXXXXXXXXXXXXXXXXdevices.net"
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Restoring device credentials from backup
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Device registration is unchanged in Azure and matches local provisioning backup.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Successful DPS provisioning.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Finished provisioning edge device.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Initializing the module runtime...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Initializing module runtime...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Using runtime network id azure-iot-edge
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Successfully initialized module runtime
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Finished initializing the module runtime.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Stopping all modules...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Finished stopping modules.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Detecting if configuration file has changed...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - No change to configuration file detected.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Edge issuer CA expiration date: 2021-03-04T10:27:22Z
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Obtaining workload CA succeeded.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Starting management API...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Starting workload API...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Starting watchdog with 60 second frequency...
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Listening on unix:///var/lib/iotedge/mgmt.sock with 1 thread for management API.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Listening on unix:///var/lib/iotedge/workload.sock with 1 thread for workload API.
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Checking edge runtime status
Dec 04 09:35:44 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:44Z [INFO] - Creating and starting edge runtime module edgeAgent
Dec 04 09:35:49 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:49Z [WARN] - Error in watchdog when checking for edge runtime status:
Dec 04 09:35:49 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:49Z [WARN] - A module runtime error occurred.
Dec 04 09:35:49 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:49Z [WARN] -         caused by: Could not get identity $edgeAgent
Dec 04 09:35:49 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:49Z [WARN] -         caused by: Could not get module $edgeAgent
Dec 04 09:35:49 wklazuedgq01.ap.cat.com iotedged[131037]: 2020-12-04T15:35:49Z [WARN] -         caused by: HTTP request failed: [401 Unauthorized] {"Message":"{\"errorCode\":401002,\"trackingId\":\"7b0967c1f33c4e80b8fb8fd2768f6814-G:16-TimeStamp:12/04/2020 15:35:49\",\"message\":\"Unauthorized\",\"timestampUtc\":\"2020-12-04T15:35:49.6309174Z\"}","ExceptionMessage":""}

```

The error here here is clearly an IoT Hub / DPS response, so I hope the tracking ID can be used to find out more.

</details>

<details>
<summary>edge-agent logs</summary>

```
N/A - fails to start
```
</details>

<details>
<summary>edge-hub logs</summary>

```
N/A - fails to start
```
</details>

## Additional Information

I have a strong hunch this is proxy configuration related, but do not have enough insight into edge to be able to debug this further.  This seems to match no error messages in other issues that I can find.  The customer is quite urgently awaiting a response as we've been digging through issues for several days now.  @veyalla pointed me here for on-call support, but please feel free to reach out to me on teams as well (`jolorich`).

I have a test environment set up to try and replicate this error.

**Comment by vipeller** [comment]
@jlorich Hi,
can you turn on debug logs for iotedged as the following:

https://docs.microsoft.com/en-us/azure/iot-edge/troubleshoot?view=iotedge-2018-06#check-the-status-of-the-iot-edge-security-manager-and-its-logs

(the part that sets the environment variables for iotedge.service)

**Comment by vipeller** [comment]
@jlorich I see this line dumped by iotedge check:

× host can connect to and perform TLS handshake with DPS endpoint - Error

do they have openssl on that machine? It is worth to try to connect using openssl s_client, using the keys/certificates they have to see if it works or throws an error, and if it does, what error

**Comment by vipeller** [comment]
@jlorich  Hello, have you managed to resolve this?

**Comment by jlorich** [comment]
@vipeller sorry for the delay in response - I did get it figured out.  It turns out they were using a traffic-inspecting proxy which terminates TLS and then opens a new connection.  With SAS it's fine, but with x509 since the connection was reestablished the client cert is stripped.

I updated the [proxy documentation](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-configure-proxy-support?view=iotedge-2018-06#working-with-traffic-inspecting-proxies0 for Edge to clarify this so hopefully others don't run into the same issue in the future.

Thanks!
